src/compiler/checker.ts

```diff
@@ -11939,8 +11941,14 @@ namespace ts {
                     return Ternary.True;
                 }
 
-                const sourceSignatures = getSignaturesOfType(source, kind);
-                const targetSignatures = getSignaturesOfType(target, kind);
+                const sourceIsJSConstructor = source.symbol && isJavaScriptConstructor(source.symbol.valueDeclaration);

```

Comment:
Are there code paths that call compareSignatures without eventually calling signaturesRelatedTo? It seems like the whole fix could reside here.